tools/setup_helpers/cmake.py

```diff
@@ -278,22 +278,35 @@ def generate(self, version, cmake_python_library, build_python, build_test, my_e
             'USE_SYSTEM_EIGEN_INSTALL': 'OFF'
         })
 
+        # Options starting with CMAKE_
+        cmake__options = {
+            'CMAKE_BUILD_TYPE': self._build_type,
+            'CMAKE_INSTALL_PREFIX': install_dir,
+            'CMAKE_C_FLAGS': cflags,
+            'CMAKE_CXX_FLAGS': cflags,
+            'CMAKE_EXE_LINKER_FLAGS': ldflags,
+            'CMAKE_SHARED_LINKER_FLAGS': ldflags,
+        }
+
+        # We set some CMAKE_* options in our Python build code instead of relying on the user's direct settings. Emit an
+        # error if the user also attempts to set these CMAKE options directly.
```

Comment:
Arguably, the fact that we are setting these flags is an antipattern (except maybe `CMAKE_INSTALL_PREFIX`, because we need it to install to a specific location?)

Reply:
Yes, I agree. I added a preparation PR #22875 to address the build type issue. Seems like emancipating the other four ('CMAKE_C_FLAGS', 'CMAKE_CXX_FLAGS', 'CMAKE_EXE_LINKER_FLAGS',       'CMAKE_SHARED_LINKER_FLAGS') requires more specific handling after we allow `CMAKE_` variables to pass down -- IMO should probably be done in a PR following up this one.